### PR TITLE
Use <? extends X> instead of <X> for most generics.

### PR DESCRIPTION
**Pull request by ST-DDT**
Associated Issue: https://github.com/SpongePowered/SpongeAPI/issues/369

This PR replaces most occurrences of generics with those that are easier to extend in subclasses.

If there are places that you don't want to add the more generic things, i can remove them there.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java [hunk 1]
```diff
@@ -61,7 +62,7 @@
 
     private final PluginManager pluginManager;
     private final SimpleDispatcher dispatcher = new SimpleDispatcher();
-    private final Multimap<PluginContainer, CommandMapping> owners = HashMultimap.create();
+    private final Multimap<PluginContainer, ImmutableCommandMapping> owners = HashMultimap.create();
```
Comment (nightpool) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: Isn't this out of score for this PR?

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/service/command/SimpleCommandService.java]: Maybe, i can remove it if it is not liked (I used a second commit for that so i can revert it if it is not liked.) , but from my point of view the SimpleCommandService is a class that will not be subclasses and thus can be manipulated to match its real contents.

Its more or less an example how the changes to the API could be used. 

I also work on the server implementation, but there are currently some unrelated issues that prevent my IDE from showing any useful compile errors.

**Comment by Zidane** [comment]
Cannot merge, please fix

**Comment by ryantheleach** [comment]
I've only just started learning about wildcard generics and they still confuse me, but do you really need <? extends interface> Wouldn't all classes implementing that interface fill <interface> even if they are subclasses of the original class that implemented the interface?

**Comment by gabizou** [comment]
> I've only just started learning about wildcard generics and they still confuse me, but do you really need <? extends interface> Wouldn't all classes implementing that interface fulfil < interface> even if they are subclasses of the original class that implemented the interface?

The issue arrises when a collection of `Foo` which extends `Bar` can't be returned as `List<Bar>` unless explicitly declared and/or casted as `Bar`.

**Comment by ryantheleach** [comment]
Yes, but that isn't the case we are dealing with in the API is it?

Don't we have a collection of objects that implement a common interface?

**Comment by gabizou** [comment]
> Don't we have a collection of objects that implement a common interface?

Yes, and this is why it's troublesome, when an implementation has `List<MyAnimal>` where `MyAnimal extends Animal`, it can't just return that list, it has to re-cast the list as `List<Animal>`.

On the other hand, there are no valid reasons to have fine tuning of these interfaces since there simply isn't one in the API, a `Human` is always a `Human`, regardless of implementation.

**Comment by gabizou** [comment]
Speaking with some more @SpongePowered/developers in IRC, we've made the comments that this PR specifically has the following:
- It does NOT remove any casting in the implementation since at compile time, nothing implements the API.
- It doesn't add to the API in any way, in fact, it makes certain parts of the API look a bit convoluted/confusing as they don't directly affect the API in any way.
- In implementations, casting is almost always required.

**Comment by ST-DDT** [comment]
I'm not sure whether we should use the sponge server impl as a reference. A CB like impk would defenitly have more casting required.
But this is up to you/the dev team.

Shall I stop working on the server side PR for this?

**Comment by Zidane** [comment]
I've talked with a few and heard both sides. I just don't see the merit in this when a cast is easily performed. As such, closing.

Thanks @ST-DDT for bringing this forward. Very good discussion and learning all around.

**Comment by Minecrell** [comment]
We should keep this in mind because it is and will be required at some places in the API, for example I could imagine this with events. I can't think of a specific example right now, but here is a small example showing where it is necessary:

```java
interface OptionalEntityEvent extends Event {
    // This needs to have the different type argument
    // so we can make it more specific in the extended event.
    // Optional<Entity> getEntity();
    Optional<? extends Entity> getEntity();
}

interface OptionalPlayerEvent extends OptionalEntityEvent {
    @Override
    Optional<Player> getEntity();
}
```

**Comment by Zidane** [comment]
@Minecrell 

Thats fine, it can be handled on a case by case basis.

**Comment by ryantheleach** [comment]
At some point you will want the API to be stable for a release. Handling it on a case-by-case basis means potentially missing some situations where this would be needed, and as far as I can see including this wouldn't hurt except inside the API itself.

To save pain later on wouldn't the change for all optionals be a good idea?

**Comment by gabizou** [comment]
> To save pain later on wouldn't the change for all optionals be a good idea?

The issue with a blanket application of all optionals is generally a bad idea. Just think, does it make sense that there would be a more fine grained implementation of `Game` made available in the API? The same applies to several of the optionals.

**Comment by ST-DDT** [comment]
For more implementation specific APIs it would defenitly makes sense.
SpongeAPI < ImplAPI < Impl

**Comment by modwizcode** [comment]
But it doesn't matter if we never want a more specific type. So Game
wouldn't ever make any sense here.

**Comment by Axiometry** [comment]
The takeaway from this is that Java Generics are terrible :cry:

**Comment by ST-DDT** [comment]
>The takeaway from this is that Java Generics are terrible :cry:

No, they are not terrible, they are logical and safe

> But it doesn't matter if we never want a more specific type. So Game wouldn't ever make any sense here.

So is this idea completely refused or will we change it for only a few occurrences?

**Pull request closed without merging**